src/core/DY.Core.Trace.SplitUtils.fst

```diff
@@ -0,0 +1,445 @@
+/// tr1 beginning after tr2.
+
+let rec trace_concat (#label_t:Type) (tr1 tr2:trace_ label_t)
+  : trace_ label_t
+  = match tr2 with
+    | Nil -> tr1
+    | Snoc hd e -> Snoc (trace_concat tr1 hd) e
+
+(* Old definition of trace subtraction --- to be removed before merge.
+   TODO
+let trace_subtract (#label_t:Type) (tr1:trace_ label_t) (tr2:trace_ label_t{tr2 <$ tr1})
+  : trace_ label_t
+  = match tr2 with
+    | Nil -> tr1
+    | _ -> let (_, _, tl) = trace_split tr1 (length tr2 - 1) in
+          tl
+*)
+
+let rec trace_subtract (#label_t:Type) (tr1:trace_ label_t) (tr2:trace_ label_t{tr2 <$ tr1})
+  : trace_ label_t
+  = if length tr1 = length tr2 then Nil
+    else begin
+      let Snoc hd e = tr1 in
+      grows_cases tr2 tr1;
+      assert(tr2 <$ hd);
+      Snoc (trace_subtract hd tr2) e
+    end
+
+let (<++>) = trace_concat
+let (<-->) = trace_subtract
+
+/// Traces form a monoid with trace concatenation as operation, and Nil as unit.
+
+let rec trace_concat_nil (#label_t:Type) (tr:trace_ label_t)
+  : Lemma
+    (ensures (
+      tr <++> Nil == tr /\
+      Nil <++> tr == tr
+    ))
+    [SMTPatOr [[SMTPat (tr <++> Nil)]; [SMTPat (Nil <++> tr)]]]
+  = match tr with
+    | Nil -> ()
+    | Snoc hd _ -> trace_concat_nil hd
+
+let rec trace_concat_assoc (#label_t:Type) (tr1 tr2 tr3:trace_ label_t)
+  : Lemma
+    (ensures ((tr1 <++> tr2) <++> tr3) == (tr1 <++> (tr2 <++> tr3)))
+  = match tr3 with
+    | Nil -> ()
+    | Snoc hd _ -> trace_concat_assoc tr1 tr2 hd
+
+let rec trace_subtract_nil (#label_t:Type) (tr:trace_ label_t)
+  : Lemma
+    (ensures (tr <--> Nil) == tr)
+    [SMTPat (tr <--> Nil)]
+  = match tr with
+    | Nil -> ()
+    | Snoc hd _ -> trace_subtract_nil hd
+  // Proves with () with initial trace subtract def
+  // TODO: Remove comment when trace subtraction is resolved
+
+let trace_subtract_snoc_left (#label_t:Type) (tr1 tr2:trace_ label_t) (e:trace_entry_ label_t)
+  : Lemma
+    (requires tr1 <$ tr2)
+    (ensures ((Snoc tr2 e) <--> tr1) == Snoc (tr2 <--> tr1) e)
+  = ()
+
+let trace_subtract_snoc_left' (#label_t:Type) (tr1 tr2:trace_ label_t)
+  : Lemma
+    (requires Snoc? tr2 /\ (let Snoc hd _ = tr2 in tr1 <$ hd))
+    (ensures (
+      let Snoc hd e = tr2 in
+      (tr2 <--> tr1) == Snoc (hd <--> tr1) e
+    ))
+  = ()
+
+let trace_concat_snoc_right (#label_t:Type) (tr1 tr2:trace_ label_t) (e:trace_entry_ label_t)
+  : Lemma
+    (ensures tr1 <++> (Snoc tr2 e) == Snoc (tr1 <++> tr2) e)
+  = ()
+
+let snoc_is_concat_singleton (#label_t:Type) (tr :trace_ label_t) (e:trace_entry_ label_t)
+  : Lemma
+    (ensures Snoc tr e == (tr <++> (Snoc Nil e)))
+  = ()
+
+let rec trace_concat_grows (#label_t:Type) (tr1 tr2 tr3:trace_ label_t)
+  : Lemma
+    (requires tr1 <$ tr2)
+    (ensures tr1 <$ (tr2 <++> tr3))
+    [SMTPat (tr1 <$ (tr2 <++> tr3)); SMTPat (tr1 <$ tr2)]
+  = match tr3 with
+    | Nil -> ()
+    | Snoc hd e -> trace_concat_grows tr1 tr2 hd
```

Comment:
We can replace this lemma with another more generic:
```suggestion
let rec trace_concat_grows (#label_t:Type) (tr1 tr2:trace_ label_t)
  : Lemma
    (tr1 <$ (tr1 <++> tr2))
    [SMTPat (tr1 <$ (tr1 <++> tr2))]
  = match tr2 with
    | Nil -> ()
    | Snoc hd e -> trace_concat_grows tr1 hd
```

Reply:
I believe these two lemmas are equivalent (using transitivity to get from the two-argument to the three-argument version), though the two argument version is arguably cleaner.